File: sdk-js/src/pathPrefixes.ts
import {
  DEFAULT_API_URL,
  DEFAULT_CORE_PATH_PREFIX,
  DEFAULT_GATEWAY_PATH_PREFIX,
  DEFAULT_USAGE_PATH_PREFIX,
  ECS_CORE_PATH_PREFIX,
  ECS_GATEWAY_PATH_PREFIX,
  ECS_USAGE_PATH_PREFIX,
} from './constants';
import { resolveBaseUrl } from './urls';

/** True for hosted Tollara API hosts (prod/PPE/branded *.api.tollara.ai). */
export function isHostedTollaraApiOrigin(origin: string): boolean {
  try {
    const host = new URL(origin).hostname.toLowerCase();
    if (host === 'api.tollara.ai' || host.endsWith('.api.tollara.ai')) {
      return true;
    }
    if (host === 'api.ppe.tollara.ai' || host.endsWith('.api.ppe.tollara.ai')) {
      return true;
    }
    return false;
  } catch {
    return false;
  }
}

export function resolveGatewayPathPrefix(
  baseUrl?: string | null,
  override?: string | null,
): string {
  const explicit = override?.trim();
  if (explicit) {
    return explicit;
  }
  const origin = resolveBaseUrl(baseUrl, DEFAULT_API_URL);
  return isHostedTollaraApiOrigin(origin) ? ECS_GATEWAY_PATH_PREFIX : DEFAULT_GATEWAY_PATH_PREFIX;
}

export function resolveCorePathPrefix(baseUrl?: string | null, override?: string | null): string {
  const explicit = override?.trim();
  if (explicit) {
    return explicit;
  }
  const origin = resolveBaseUrl(baseUrl, DEFAULT_API_URL);
  return isHostedTollaraApiOrigin(origin) ? ECS_CORE_PATH_PREFIX : DEFAULT_CORE_PATH_PREFIX;
}

export function resolveUsagePathPrefix(baseUrl?: string | null, override?: string | null): string {
  const explicit = override?.trim();
  if (explicit) {
    return explicit;
  }
  const origin = resolveBaseUrl(baseUrl, DEFAULT_API_URL);
  return isHostedTollaraApiOrigin(origin) ? ECS_USAGE_PATH_PREFIX : DEFAULT_USAGE_PATH_PREFIX;
}

File: sdk-go/path_prefixes.go
package sdk

import (
	"net/url"
	"strings"
)

const (
	EcsCorePathPrefix    = "/core/api/v1"
	EcsGatewayPathPrefix = "/gateway/api/v1"
	EcsUsagePathPrefix   = "/usage/api/v1"
)

// IsHostedTollaraApiOrigin reports whether origin is hosted Tollara (prod/PPE/branded *.api.tollara.ai).
func IsHostedTollaraApiOrigin(origin string) bool {
	origin = strings.TrimSpace(origin)
	if origin == "" {
		return false
	}
	u, err := url.Parse(origin)
	if err != nil || u.Host == "" {
		return false
	}
	host := strings.ToLower(u.Hostname())
	if host == "api.tollara.ai" || strings.HasSuffix(host, ".api.tollara.ai") {
		return true
	}
	return host == "api.ppe.tollara.ai" || strings.HasSuffix(host, ".api.ppe.tollara.ai")
}

func resolveOrigin(baseURL string) string {
	trimmed := strings.TrimSpace(baseURL)
	if trimmed != "" {
		return trimTrailingSlash(trimmed)
	}
	return DefaultAPIURL
}

// ResolveGatewayPathPrefix selects ECS vs Docker gateway prefix from base URL (explicit override wins).
func ResolveGatewayPathPrefix(baseURL, override string) string {
	if strings.TrimSpace(override) != "" {
		return strings.TrimSpace(override)
	}
	origin := resolveOrigin(baseURL)
	if IsHostedTollaraApiOrigin(origin) {
		return EcsGatewayPathPrefix
	}
	return DefaultGatewayPathPrefix
}

// ResolveCorePathPrefix selects ECS vs Docker core prefix from base URL (explicit override wins).
func ResolveCorePathPrefix(baseURL, override string) string {
	if strings.TrimSpace(override) != "" {
		return strings.TrimSpace(override)
	}
	origin := resolveOrigin(baseURL)
	if IsHostedTollaraApiOrigin(origin) {
		return EcsCorePathPrefix
	}
	return DefaultCorePathPrefix
}

// ResolveUsagePathPrefix selects ECS vs Docker usage prefix from base URL (explicit override wins).
func ResolveUsagePathPrefix(baseURL, override string) string {
	if strings.TrimSpace(override) != "" {
		return strings.TrimSpace(override)
	}
	origin := resolveOrigin(baseURL)
	if IsHostedTollaraApiOrigin(origin) {
		return EcsUsagePathPrefix
	}
	return DefaultUsagePathPrefix
}
